File: src/ios/PoilabsAnalysisModule.m
#import <Foundation/Foundation.h>
#import "PoilabsAnalysisModule.h"

@implementation PoilabsAnalysisModule

RCT_EXPORT_MODULE(PoilabsAnalysisModule);

RCT_EXPORT_METHOD(startPoilabsAnalysis:(NSString *)applicationId
                  applicationSecret:(NSString *)secret
                  uniqueIdentifier:(NSString *)uniqueId
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  [[PLAnalysisSettings sharedInstance] setApplicationId:applicationId];
  [[PLAnalysisSettings sharedInstance] setApplicationSecret:secret];
  [[PLAnalysisSettings sharedInstance] setAnalysisUniqueIdentifier:uniqueId];

  __block BOOL promiseSettled = NO; 

  [[PLConfigManager sharedInstance] getReadyForTrackingWithCompletionHandler:^(PLError *error) {
    if (promiseSettled) {
      return; 
    }
    promiseSettled = YES;

    if (error) {
      NSLog(@"Poilabs init error: %@", error.errorDescription);
      reject(@"PoilabsInitError", error.errorDescription, error);
    } else {
      [[PLSuspendedAnalysisManager sharedInstance] stopBeaconMonitoring];
      [[PLStandardAnalysisManager sharedInstance] startBeaconMonitoring];
      [[PLStandardAnalysisManager sharedInstance] setDelegate:self];
      resolve(@YES);
    }
  }];
}

RCT_EXPORT_METHOD(stopPoilabsAnalysis:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  [[PLAnalysisSettings sharedInstance] closeAllActions];
  resolve(@YES);
}

RCT_EXPORT_METHOD(updateUniqueId:(NSString *)uniqueId
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  [[PLAnalysisSettings sharedInstance] setAnalysisUniqueIdentifier:uniqueId];
  resolve(@YES);
}

#pragma mark – PLAnalysisManagerDelegate

- (void)analysisManagerDidFailWithPoiError:(PLError *)error
{
  NSLog(@"Poilabs error: %@", error);
}

- (void)analysisManagerResponseForBeaconMonitoring:(NSDictionary *)response
{
  NSLog(@"Poilabs response: %@", response);
}

@end

File: src/ios/PoilabsAnalysisModule.swift
import PoilabsAnalysis
import React

@objc(PoilabsAnalysisModule)
class PoilabsAnalysisModule: NSObject {
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  @objc
  func startPoilabsAnalysis(_ applicationId: String, applicationSecret: String, uniqueId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    var promiseResolved = false
    
    DispatchQueue.main.async {
      PLAnalysisSettings.sharedInstance().applicationId = applicationId
      PLAnalysisSettings.sharedInstance().applicationSecret = applicationSecret
      PLAnalysisSettings.sharedInstance().analysisUniqueIdentifier = uniqueId
      
      PLConfigManager.sharedInstance().getReadyForTracking { error in
        if !promiseResolved {
          promiseResolved = true
          
          if let error = error {
            print("Poilabs Error: \(error)")
            resolver(false)
          } else {
            print("Poilabs initialized successfully")
            PLSuspendedAnalysisManager.sharedInstance()?.stopBeaconMonitoring()
            PLStandardAnalysisManager.sharedInstance()?.startBeaconMonitoring()
            PLStandardAnalysisManager.sharedInstance().delegate = PoilabsAnalysisDelegate.shared
            resolver(true)
          }
        }
      }
    }
  }
    
  @objc
  func stopPoilabsAnalysis(_ resolve: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      PLAnalysisSettings.sharedInstance().closeAllActions()
      resolve(true)
    }
  }
  
  @objc
  func updateUniqueId(_ uniqueId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      PLAnalysisSettings.sharedInstance().analysisUniqueIdentifier = uniqueId
      resolver(true)
    }
  }
}